torch/csrc/jit/tensorexpr/loopnest.cpp

```diff
@@ -1779,6 +1779,135 @@ bool LoopNest::flatten(const std::vector<For*>& loops, For** flattened) {
+  // variables to access the given buffer could be optimized away.
+  //
+  // Algorithm:
+  //   * Find all the accesses to the given buf. (A)
+  //   * Find the parent common to all accesses in A. (P)
+  //   * Collect all the loops above P. (L)
+  //   * Collect all the loop variables corresponding to L. (LV)
+  //   * For every access a in A:
+  //      * For the index I in every dimension of a:
+  //          * If the variables in I are all in LV, mark this dimension
+  //            for deletion.
+  //   * For every dimension that is marked for deletion in ALL accesses in A:
+  //      * Update the buffer to set the size of that dimension to 1.
+  //      * Update all accesses in A to set the index in that dimension to 0.
+
+  auto writes = WritesToBuf::find(stmt, buf);
+  auto reads = StmtsReadingBuf::find(stmt, buf);
+
+  // All buffers must be read and written at least once.
+  // Is this a valid assumption? TODO
+  TORCH_INTERNAL_ASSERT(!writes.empty());
+  TORCH_INTERNAL_ASSERT(!reads.empty());
+
+  // Find the parent common to all the buffer accesses.
+  const Block* parent = dynamic_cast<Block*>(writes.front()->get_parent());
+  TORCH_INTERNAL_ASSERT(parent);
+  for (auto w : writes) {
+    parent = Block::getSharedParent(parent, w);
+  }
+  for (auto r : reads) {
+    parent = Block::getSharedParent(parent, r);
+  }
+  std::cout << "common parent: " << *parent << std::endl;
+
+  // Collect all the loops that are above the common parent.
+  auto loops = LoopNest::getEnclosingLoopNest(parent);
+  std::unordered_set<const Var*> loop_vars;
+  for (auto l : loops) {
+    loop_vars.insert(l->var());
+    std::cout << "enclosing loop: " << *l << std::endl;
+  }
+
+  // TODO: Need to handle other Stmts / Exprs that read / write buffers.
+  auto stores = NodeFinder<Store>::find(stmt);
+  auto loads = NodeFinder<Load>::find(stmt);
+
+  // Vector to indicate which dimensions could be compressed away.
+  std::vector<bool> dims(buf->dims().size(), true);
+  auto check_indices = [&](const std::vector<const Expr*>& indices) {
+    TORCH_INTERNAL_ASSERT(indices.size() == dims.size());
+    for (size_t i = 0; i < indices.size(); ++i) {
+      auto index_vars = NodeFinder<Var>::find(indices[i]);
+      for (auto iv : index_vars) {
+        if (loop_vars.count(iv) == 0) {
+          // A variable in this index is not in loop_vars.
+          // This implies that this dimension cannot be optimized away.
+          dims[i] = false;
+          break;
+        }
+      }
+    }
+  };
+  for (auto s : stores) {
+    if (s->buf() == buf) {
+      check_indices(s->indices());
+    }
+  }
+  for (auto l : loads) {
+    if (l->buf() == buf) {
+      check_indices(l->indices());
+    }
+  }
+  int num_dims_to_compress = 0;
+  for (auto d : dims) {
+    std::cout << "dim: " << d << std::endl;
+    num_dims_to_compress += static_cast<int>(d);
+  }
+  if (num_dims_to_compress == 0) {
+    return;
+  }
```

Comment:
This probably would be more elegant if we have a `bool` flag `any_dim_to_compress`. We won't need a `bool` to `int` cast then.

Reply:
Good point. Done.